google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/JsonToProtoMessage.java

```diff
@@ -0,0 +1,406 @@
+ */
+package com.google.cloud.bigquery.storage.v1alpha2;
+
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.DynamicMessage;
+import com.google.protobuf.Message;
+import java.util.HashMap;
+import java.util.List;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+/** Converts Json data to protocol buffer messages given the protocol buffer descriptor. */
+public class JsonToProtoMessage {
+
+  /**
+   * Converts Json data to protocol buffer messages given the protocol buffer descriptor.
+   *
+   * @param protoSchema
+   * @param json
+   * @throws IllegalArgumentException when JSON data is not compatible with proto descriptor.
+   */
+  public static DynamicMessage convertJsonToProtoMessage(Descriptor protoSchema, JSONObject json)
+      throws IllegalArgumentException {
+    return convertJsonToProtoMessage(protoSchema, json, false);
+  }
+
+  /**
+   * Converts Json data to protocol buffer messages given the protocol buffer descriptor.
+   *
+   * @param protoSchema
+   * @param json
+   * @param allowUnknownFields Ignores unknown JSON fields.
+   * @throws IllegalArgumentException when JSON data is not compatible with proto descriptor.
+   */
+  public static DynamicMessage convertJsonToProtoMessage(
+      Descriptor protoSchema, JSONObject json, boolean allowUnknownFields)
+      throws IllegalArgumentException {
+    return convertJsonToProtoMessageImpl(protoSchema, json, "root", true, allowUnknownFields);
+  }
+
+  /**
+   * Converts Json data to protocol buffer messages given the protocol buffer descriptor.
+   *
+   * @param protoSchema
+   * @param json
+   * @param jsonScope Debugging purposes
+   * @param topLevel If root level has no matching fields, throws exception.
+   * @param allowUnknownFields Ignores unknown JSON fields.
+   * @throws IllegalArgumentException when JSON data is not compatible with proto descriptor.
+   */
+  private static DynamicMessage convertJsonToProtoMessageImpl(
+      Descriptor protoSchema,
+      JSONObject json,
+      String jsonScope,
+      boolean topLevel,
+      boolean allowUnknownFields)
+      throws IllegalArgumentException {
+    DynamicMessage.Builder protoMsg = DynamicMessage.newBuilder(protoSchema);
+    int matchedFields = 0;
+    List<FieldDescriptor> protoFields = protoSchema.getFields();
+
+    if (JSONObject.getNames(json).length > protoFields.size()) {
+      if (!allowUnknownFields) {
+        throw new IllegalArgumentException(
+            "JSONObject has fields unknown to BigQuery: f1. Set allowUnknownFields to True to allow unknown fields.");
+      }
+    }
+    HashMap<String, String> jsonLowercaseNameToActualName = new HashMap<String, String>();
+    String[] actualNames = JSONObject.getNames(json);
+    for (int i = 0; i < actualNames.length; i++) {
+      jsonLowercaseNameToActualName.put(actualNames[i].toLowerCase(), actualNames[i]);
```

Comment:
Instead of creating a new container, maybe use case insenstive set comparison directly?
https://stackoverflow.com/questions/24558456/java-setstring-equality-ignore-case

Reply:
So I think I still need the HashMap to hold a mapping from <LowercasedName, Name> because in order retrieve the vjson alue, I still need to use the original name since I need to call JSONObject.get(<original name>).

Reply:
Can it be:
 Set<String> jsonNames = new TreeSet<String>(String.CASE_INSENSITIVE_ORDER);
 jsonNames.addAll(JSONObject.getName(json));

...
jsonNames.contains(field.getName())

?

Reply:
Yes, and this does check if the field names are case-insensitive equal. But after checking, I need to know the actual name of the json field in order to access the json value, so I think I need to create a mapping from <LowercasedName -> Name>